### PR TITLE
[docs] Start a doc on ClangImporter C->Swift name mapping

### PR DESCRIPTION
**Pull request by jrose-apple**
Includes enums in all their flavors, plus `swift_wrapper` structs.

Planned future content:

- Methods, initializers, properties
  - "Omit-needless-words" rules
  - NSError
  - Boolean properties
  - Inferred default arguments
  - Factory methods as initializers
- Protocols (mainly just that they check for collision)
- CF types (low-priority since non-Apple people can't make them)
- Custom names, including import-as-member and import-as-accessor
- swift_private

### CODE REVIEW

**Comment by jrose-apple** [comment]
@swift-ci Please smoke test

File: docs/CImportNameTranslation.md [hunk 1]
```diff
@@ -0,0 +1,191 @@
...
+
+  static var badness: VagueFailure.Code { get }
+  static var worseness: VagueFailure.Code { get }
+  static var worstness: VagueFailure.Code { get }
+
+  static var errorDomain: String { get }
+}
+```
+
+- Struct name is enum name, optionally dropping the word "Code".
```
Comment (xymus) [inline docs/CImportNameTranslation.md]: Is this optional to Omit-needless-words?

**Review by xymus — APPROVED** [review]

File: docs/CImportNameTranslation.md (hunk 1 above)
Comment (jrose-apple) [reply docs/CImportNameTranslation.md]: Whoops. I should probably just say "dropping the suffix `"Code"` if present".

File: docs/CImportNameTranslation.md [hunk 2]
```diff
@@ -0,0 +1,191 @@
+This document gives a high-level description of how C and Objective-C declarations are translated to Swift, with particular focus on how names are adjusted. It is not attempting to be a *complete* description of everything the compiler does except with regards to how *names* are transformed; even there, some special cases that only apply to Apple's SDKs have been omitted.
```
Comment (brentdax) [inline docs/CImportNameTranslation.md]: This should probably have a title.

**Review by beccadax — CHANGES_REQUESTED** [review]
This is a very welcome addition to our docs folder!

File: docs/CImportNameTranslation.md (hunk 2 above)
Comment (jrose-apple) [reply docs/CImportNameTranslation.md]: I'm always conflicted about that because the filename is sort of a title, but sure.

**Review by beccadax — APPROVED** [review]

**Review by nkcsgexi — APPROVED** [review]
Thanks for adding this! Could you mention where the name translation APIs are declared in the compiler source? I presume it's `SwiftNameTranslation.h`?

**Comment by jrose-apple** [comment]
swift/AST/SwiftNameTranslation.h is for Swift-to-ObjC, not ObjC-to-Swift. Most of what's talked about here is in swift/ClangImporter/ImportName.h, but not all of it.

**Comment by jrose-apple** [comment]
@swift-ci Please smoke test

**Comment by nkcsgexi** [comment]
ah, OK. Notice that both directions of translation are interesting for cross-language automatic renaming.

**Comment by jrose-apple** [comment]
> Notice that both directions of translation are interesting for cross-language automatic renaming.

I think I'll rename the file to CToSwiftNameTranslation, leaving room for a future SwiftToCNameTranslation.

**Comment by nkcsgexi** [comment]
> I think I'll rename the file to CToSwiftNameTranslation, leaving room for a future SwiftToCNameTranslation.

Sounds good, Thank you!

**Comment by jrose-apple** [comment]
@swift-ci Please smoke test

**Comment by jrose-apple** [comment]
We've got three approving reviewers so far, so if anyone else has comments we can take it in post-commit review.

**Pull request merged**